### PR TITLE
Drop redundant project id null checks.

### PR DESCRIPTION
**Pull request by jmcarp**
The `fallback_to_default_project_id` method already raises an exception
if both `self.project_id` and `kwargs["project_id"]` are missing, so
skip redundant null checks in hook methods.

---
Make sure to mark the boxes below before creating PR: [x]

- [x] Description above provides context of the change
- [x] Unit tests coverage for changes (not needed for documentation changes)
- [x] Commits follow "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)"
- [x] Relevant documentation is updated including usage instructions.
- [x] I will engage committers as explained in [Contribution Workflow Example](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#contribution-workflow-example).

---
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).
Read the [Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines) for more information.

### CODE REVIEW

**Comment by turbaszek** [comment]
@jmcarp those checks are only to please mypy. Initially, we used `assert project_id` but since we don't use `assert` in production code we have if clause...

**Comment by potiuk** [comment]
Just thinking - maybe we can do it differently and find a way to both - keep mypy happy and remove the ifs... Let me think about it a bit :)

**Comment by turbaszek** [comment]
@potiuk I think @mik-laj already gave it a lot of thought

**Comment by potiuk** [comment]
Some more would not hurt :)

**Comment by turbaszek** [comment]
Sure, would be awesome to remove it. But typehints with decorators are tricky :)

**Comment by potiuk** [comment]
Hold my  :beer: and :cat:

**Comment by turbaszek** [comment]
> Hold my 🍺 and 🐱

Are you going to propose PEP to make Python statically typed language? 😅

**Comment by potiuk** [comment]
Yep. That's my secret plan. Just started to write it :)

**Comment by jmcarp** [comment]
I added a simple mypy plugin to preserve types for decorated functions, and to update types when decorators change function signatures. For example, `fallback_to_default_project_id` effectively changes the type of the `project_id` argument from `str` to `Optional[str]`. This also fixes the problem of decorators changing the types of all decorated callables to `Callable[..., Any]`. Still a WIP, but what do you think?

**Comment by turbaszek** [comment]
I have mixed feelings... First, I am not a big fan of adding more and more custom plugins (especially to solve problems that should be solved on language level #static_types_fan) . Especially to solve problems that I am not even sure are real problems. Since we have the `if` clause I even started to like the "explicit" information about `project_id` ("at this point, it will be provided by decorator").  

On the other hand, this change does no harm. It makes our codebase bigger and adds next moving part to it. So, I am not sure.

I am curious about what others think? @mik-laj @kaxil @ashb @potiuk

**Comment by ashb** [comment]
Quick thought: The same issue about project_id also applies to session/provide_session

**Comment by jmcarp** [comment]
I was also hesitant about writing a mypy plugin, even a simple one, but as far as I can tell, it's the only way to write decorators that don't produce callables with type `Callable[..., Any]`. I also think it's not ideal that all decorated functions in the airflow codebase have that type. For example, users might see that `PubSubHook.publish` has type annotations and expect that mypy will catch type errors. But that's not the case, since the `fallback_to_default_project_id` decorator doesn't preserve function types. Adding the mypy plugin also caused the type checker to find some type errors in the codebase that it couldn't catch previously, again because decorated functions are effectively untyped.

**Comment by jmcarp** [comment]
One more thing: one of the mypy developers suggests using mypy plugins to preserve types for decorated functions until mypy supports this kind of thing: https://github.com/python/mypy/issues/1927#issuecomment-305152052.

**Comment by turbaszek** [comment]
> Quick thought: The same issue about project_id also applies to session/provide_session

As well as any other decorator that will inject arguments. @jmcarp will we be able to somehow check if all decorators that require it use this plugin? `project_id` is only GCP stuff an probably not so much problematic but `provide_session` is common and its case can be worth considering.

**Comment by jmcarp** [comment]
I don't know if we can programmatically audit use of the plugin, but the number of decorators is small and shouldn't change often, and it's simple enough to check by hand by grepping for `wraps`. I can add more decorators to the plugin here, but this patch is already pretty big, so I was thinking of adding `provide_session` and other decorators in a follow-up if we accept this change.

File: airflow/providers/google/cloud/example_dags/example_automl_tables.py [hunk 1]
```diff
@@ -139,7 +139,6 @@ def get_target_column_spec(columns_specs: List[Dict], column_name: str) -> str:
         task_id="update_dataset_task",
         dataset=update,
         location=GCP_AUTOML_LOCATION,
-        project_id=GCP_PROJECT_ID,
```
Comment (turbaszek) [inline airflow/providers/google/cloud/example_dags/example_automl_tables.py:142]: Is this change necessary?

File: airflow/providers/google/cloud/hooks/bigtable.py [hunk 2]
```diff
@@ -54,7 +54,7 @@ def _get_client(self, project_id: str):
         return self._client
 
     @GoogleBaseHook.fallback_to_default_project_id
-    def get_instance(self, instance_id: str, project_id: Optional[str] = None) -> Instance:
+    def get_instance(self, instance_id: str, project_id: str) -> Instance:
```
Comment (turbaszek) [inline airflow/providers/google/cloud/hooks/bigtable.py:57]: Let's decide what we want to use: `str` or `Optional[str] = None` and be cosistent

File: tests/providers/google/cloud/hooks/test_cloud_sql.py [hunk 3]
```diff
@@ -561,29 +513,6 @@ def test_get_instance_overridden_project_id(
...
-        wait_for_operation_to_complete.return_value = None
-        with self.assertRaises(AirflowException) as cm:
-            self.cloudsql_hook_no_default_project_id.get_instance(
-                instance='instance')
-        get_method.assert_not_called()
-        execute_method.assert_not_called()
-        err = cm.exception
-        self.assertIn("The project id must be passed", str(err))
-        wait_for_operation_to_complete.assert_not_called()
-
```
Comment (turbaszek) [inline tests/providers/google/cloud/hooks/test_cloud_sql.py:586]: Why remove those tests?

File: airflow/providers/google/cloud/example_dags/example_automl_tables.py (hunk 1 above)
Comment (potiuk) [reply airflow/providers/google/cloud/example_dags/example_automl_tables.py:142]: Yeah. I think we should leave that in.

File: airflow/mypy/plugin/decorators.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
+#
```
Comment (potiuk) [inline airflow/mypy/plugin/decorators.py:1]: I like this. I think we should open-up to more custom plugins for mypy/pylint. We'd learn how to develop them and this will help us when we will do automated checks for Dags to see if they are 2.0 compliant. I think it's good to have that in our toolbox.

File: airflow/providers/google/cloud/hooks/automl.py [hunk 5]
```diff
@@ -499,9 +481,6 @@ def update_dataset(
         :param update_mask: The update mask applies to the resource.  If a dict is provided, it must
             be of the same form as the protobuf message FieldMask.
         :type update_mask: Union[dict, FieldMask]
-        :param project_id: ID of the Google Cloud project where dataset is located if None then
```
Comment (potiuk) [inline airflow/providers/google/cloud/hooks/automl.py:502]: While this is correct (seems that update_dataset does not need project_id) I think it should have an entry in UPDATING.md ?

File: airflow/providers/google/cloud/hooks/bigtable.py (hunk 2 above)
Comment (potiuk) [reply airflow/providers/google/cloud/hooks/bigtable.py:57]: Yeah. Agree. project_id: str is better IMHO. it is semantically correct - we expect project_id to be set. And  it is used most of the time below.

File: airflow/providers/google/cloud/hooks/datacatalog.py [hunk 6]
```diff
@@ -659,7 +659,7 @@ def list_tags(
         location: str,
         entry_group: str,
         entry: str,
-        page_size: int = 100,
+        page_size: Optional[int] = 100,
```
Comment (potiuk) [inline airflow/providers/google/cloud/hooks/datacatalog.py:663]: Is it optional really ? I think it's quite ok to set it as mandatory since it has default

File: airflow/providers/google/cloud/hooks/dataflow.py [hunk 7]
```diff
@@ -585,10 +579,10 @@ def start_python_dataflow(  # pylint: disable=too-many-arguments
         variables: Dict,
         dataflow: str,
         py_options: List[str],
+        project_id: str,
```
Comment (potiuk) [inline airflow/providers/google/cloud/hooks/dataflow.py:582]: I think it needs an entry in UPDATING.md. If someone uses positional args - it will break.

File: airflow/providers/google/cloud/hooks/dataflow.py [hunk 8]
```diff
@@ -730,10 +718,10 @@ def is_job_dataflow_running(self, name: str, variables: Dict, project_id: Option
     @GoogleBaseHook.fallback_to_default_project_id
     def cancel_job(
         self,
+        project_id: str,
```
Comment (potiuk) [inline airflow/providers/google/cloud/hooks/dataflow.py:721]: Same here -> I think it needs UPDATING.md entry

File: airflow/providers/google/cloud/hooks/stackdriver.py [hunk 9]
```diff
@@ -186,10 +186,10 @@ def enable_alert_policies(
     @GoogleBaseHook.fallback_to_default_project_id
     def disable_alert_policies(
         self,
-        project_id: Optional[str] = None,
+        project_id: str,
         filter_: Optional[str] = None,
         retry: Optional[str] = DEFAULT,
-        timeout: Optional[str] = DEFAULT,
+        timeout: Optional[float] = DEFAULT,
```
Comment (potiuk) [inline airflow/providers/google/cloud/hooks/stackdriver.py:192]: Nice!

File: airflow/providers/google/cloud/hooks/stackdriver.py [hunk 10]
```diff
@@ -433,8 +433,8 @@ def list_notification_channels(
     @GoogleBaseHook.fallback_to_default_project_id
     def _toggle_channel_status(
         self,
-        new_state: str,
-        project_id: Optional[str] = None,
+        new_state: bool,
```
Comment (potiuk) [inline airflow/providers/google/cloud/hooks/stackdriver.py:436]: Nice too !

File: airflow/providers/google/cloud/hooks/stackdriver.py [hunk 11]
```diff
@@ -500,8 +500,8 @@ def enable_notification_channels(
     @GoogleBaseHook.fallback_to_default_project_id
     def disable_notification_channels(
         self,
+        project_id: str,
```
Comment (potiuk) [inline airflow/providers/google/cloud/hooks/stackdriver.py:503]: UPDATING.md entry needed here as well.

File: airflow/providers/google/cloud/operators/automl.py [hunk 12]
```diff
@@ -568,7 +568,6 @@ def __init__(
         self,
         dataset: dict,
         location: str,
-        project_id: Optional[str] = None,
```
Comment (potiuk) [inline airflow/providers/google/cloud/operators/automl.py:571]: Should be removed from the docstring and mentioned in UPDATING.md

File: airflow/providers/google/cloud/operators/dataflow.py [hunk 13]
```diff
@@ -172,6 +172,7 @@ def __init__(
             job_name: str = '{{task.task_id}}',
             dataflow_default_options: Optional[dict] = None,
             options: Optional[dict] = None,
+            project_id: Optional[str] = None,
```
Comment (potiuk) [inline airflow/providers/google/cloud/operators/dataflow.py:175]: Nice! But needs UPDATING.md entry

File: airflow/providers/google/cloud/operators/dataflow.py [hunk 14]
```diff
@@ -437,6 +443,7 @@ def __init__(  # pylint: disable=too-many-arguments
             py_options: Optional[List[str]] = None,
             py_requirements: Optional[List[str]] = None,
             py_system_site_packages: bool = False,
+            project_id: Optional[str] = None,
```
Comment (potiuk) [inline airflow/providers/google/cloud/operators/dataflow.py:446]: Needs UPDATING.md entry

File: tests/providers/google/cloud/hooks/test_cloud_sql.py (hunk 3 above)
Comment (potiuk) [reply tests/providers/google/cloud/hooks/test_cloud_sql.py:586]: I think it makes sense to remove them  - they are all testing the missing project_id which is not really possible in reality.

**Review by potiuk — CHANGES_REQUESTED** [review]
Few small thigs. I like it :)

File: airflow/providers/google/cloud/hooks/automl.py (hunk 5 above)
Comment (turbaszek) [reply airflow/providers/google/cloud/hooks/automl.py:502]: @potiuk no need I think, automl is master only

File: airflow/providers/google/cloud/hooks/stackdriver.py (hunk 11 above)
Comment (turbaszek) [reply airflow/providers/google/cloud/hooks/stackdriver.py:503]: @potiuk was this hook released?

File: airflow/providers/google/cloud/example_dags/example_automl_tables.py (hunk 1 above)
Comment (jmcarp) [reply airflow/providers/google/cloud/example_dags/example_automl_tables.py:142]: This parameter was unused in `CloudAutoMLHook.update_dataset`, so I dropped the parameter from that method, as well as `AutoMLTablesUpdateDatasetOperator` and the example code here.

File: airflow/providers/google/cloud/hooks/datacatalog.py (hunk 6 above)
Comment (jmcarp) [reply airflow/providers/google/cloud/hooks/datacatalog.py:663]: We mark this as `Optional[int]` in `CloudDataCatalogListTagsOperator`, so if we don't also mark it as `Optional[int]` here, mypy errors:

```
airflow/providers/google/cloud/operators/datacatalog.py:1287: error: Argument "page_size" to "list_tags" of "CloudDataCatalogHook" has incompatible type "Optional[int]"; expected "int"
```

This is the kind of type error that mypy wasn't catching before because the decorator on the hook method dropped type information about the method, by the way.

File: tests/providers/google/cloud/hooks/test_cloud_sql.py (hunk 3 above)
Comment (mik-laj) [reply tests/providers/google/cloud/hooks/test_cloud_sql.py:586]: It's possible in reality. For example: If you use ``gcloud`` for authorization but you do not set the ``core/project`` option in gcloud, project_id will be empty. It is not guaranteed that every service account key must have project_id.  Keys that were generated a long time ago do not have this field.

File: airflow/providers/google/cloud/hooks/dataflow.py (hunk 7 above)
Comment (mik-laj) [reply airflow/providers/google/cloud/hooks/dataflow.py:582]: This method must be called using keywords arguments. Please look at this expectation added by @potiuk  https://github.com/apache/airflow/blob/master/airflow/providers/google/common/hooks/base_google.py#L333-L336

File: airflow/providers/google/cloud/hooks/dataflow.py (hunk 8 above)
Comment (mik-laj) [reply airflow/providers/google/cloud/hooks/dataflow.py:721]: It is the only keyword argument method.

**Comment by mik-laj** [comment]
If you would like to add support for more decorators, the most popular decorator is `apply_defaullts. However, this will be best done in a separate PR. This is very problematic and has made it difficult for me to introduce metaclasses - https://github.com/apache/airflow/pull/7450

File: tests/providers/google/cloud/hooks/test_cloud_sql.py (hunk 3 above)
Comment (jmcarp) [reply tests/providers/google/cloud/hooks/test_cloud_sql.py:586]: We have a large number of similar tests that all basically test the `fallback_to_default_project_id` decorator. I think it's sufficient to test the decorator on its own--I don't think we need to test every method that it's applied to.

File: airflow/providers/google/cloud/hooks/automl.py (hunk 5 above)
Comment (potiuk) [reply airflow/providers/google/cloud/hooks/automl.py:502]: Right !

File: airflow/providers/google/cloud/hooks/datacatalog.py (hunk 6 above)
Comment (potiuk) [reply airflow/providers/google/cloud/hooks/datacatalog.py:663]: OK. Making MyPy happy :)

File: airflow/providers/google/cloud/hooks/dataflow.py (hunk 7 above)
Comment (potiuk) [reply airflow/providers/google/cloud/hooks/dataflow.py:582]: Ah.. Right :)

Comment (potiuk) [reply airflow/providers/google/cloud/hooks/dataflow.py:582]: Totally forgot about it ! It was some 1.5 year ago! Thanks @mik-laj for reminding :)

File: tests/providers/google/cloud/hooks/test_cloud_sql.py (hunk 3 above)
Comment (potiuk) [reply tests/providers/google/cloud/hooks/test_cloud_sql.py:586]: Agree. No need to repeat it.

**Review by potiuk — APPROVED** [review]

File: airflow/providers/google/cloud/example_dags/example_automl_tables.py (hunk 1 above)
Comment (potiuk) [reply airflow/providers/google/cloud/example_dags/example_automl_tables.py:142]: Yeah. I thik it's good :)

**Comment by potiuk** [comment]
@turbaszek @mik-laj -> looks good to me. Can you take another look?

File: airflow/providers/google/cloud/hooks/bigtable.py (hunk 2 above)
Comment (turbaszek) [reply airflow/providers/google/cloud/hooks/bigtable.py:57]: @jmcarp I checked the changes and in some places where we use the decorator we are using `project_id: str` and in others `project_id: Optional[str] = None`. I think we should be consistent :)

**Comment by turbaszek** [comment]
I still have concerns about type annotations consistency:
https://github.com/apache/airflow/pull/8145#discussion_r407044132

**Review by turbaszek — APPROVED** [review]
LGTM

**Comment by jmcarp** [comment]
Thanks for reviewing. Is it all right to merge while `TestCliTaskBackfill.test_run_ignores_all_dependencies` is failing? It looks unrelated to this patch.

**Comment by potiuk** [comment]
Thanks @jmcarp! Great change!

**Comment by zachliu** [comment]
@jmcarp do you plan to work on `apply_defaults`? :grin:
